README.md
# request-as-curl
[![NPM Version](https://badge.fury.io/js/request-as-curl.png)]
(https://npmjs.org/package/request-as-curl)
[![Build Status](https://travis-ci.org/azproduction/node-request-as-curl.png?branch=master)]
(https://travis-ci.org/azproduction/node-request-as-curl)
[![Coverage Status](https://coveralls.io/repos/azproduction/node-request-as-curl/badge.png?branch=master)]
(https://coveralls.io/r/azproduction/node-request-as-curl)
[![Dependency Status](https://gemnasium.com/azproduction/node-request-as-curl.png)]
(https://gemnasium.com/azproduction/node-request-as-curl)

It serializes http.ClientRequest as curl(1) command string. This is like [Copy as cURL](https://twitter.com/ChromiumDev/status/317183238026186752), but for node.

## Installation

`request-as-curl` can be installed using `npm`:

```
npm install request-as-curl
```

## Example

```js
var curlify = require('request-as-curl'),
    request = require('express'),
    data = {data: 'data'};

var req = request('http://google.com/', {method: 'POST', json: data}, function (error, response, expected) {
    // Note: req.req!
    curlify(req.req, data);
    // curl 'http://google.com' -H 'accept: application/json' -H 'content-type: application/json' -H 'connection: keep-alive' --data '{"data":"data"}' --compressed
});
```

```js
var curlify = require('request-as-curl'),
    app = require('express')();

app.get('/', function (req) {
    curlify(req);
    // curl 'http://localhost/pewpew' -H 'x-real-ip: 127.0.0.1' -H 'x-forwarded-for: 127.0.0.1' -H 'x-nginx-proxy: true' -H 'connection: close' -H 'user-agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10.8; rv:25.0) Gecko/20100101 Firefox/25.0' -H 'accept: text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8' -H 'accept-language: ru-RU,ru;q=0.8,en-US;q=0.5,en;q=0.3' -H 'accept-encoding: gzip, deflate' -H 'cache-control: max-age=0' --compressed
});

app.listen();
```


CHANGELOG.md
Version 0.1.0:
 * Initial release.


test/test.request-as-curl.js
/*global describe, it, beforeEach, afterEach*/
/*jshint expr:true, maxstatements:30 */

var serializeRequest = require('..'),
    express = require('express'),
    stringify = require('json-stringify-safe'),
    http = require('http'),
    url = require('url'),
    request = require('request'),
    expect = require('chai').expect;

var PORT = 54321;
var HOSTNAME = 'localhost';
var PATH = '/p/a/t/h?query=string';
var BODY = 'data\ndata\ndata\nдата-дада●★☆♣♥✌\x01\x17';
var BODY_JSON = {
    pewpew: {
        ololo: []
    }
};

var httpRequest = function (options, content, cb) {
    var req = http.request(options, function (res) {
        res.setEncoding('utf8');

        var body = '';
        res.on('data', function (chunk) {
            body += chunk;
        });
        res.on('end', function () {
            cb(null, res, body);
        });
    });

    req.on('error', cb);

    if (content) {
        req.write(content);
    }
    req.end();

    return req;
};

describe('request-as-curl', function () {
    var app, server;
    var oldPlatform = process.platform;

    beforeEach(function(done) {
        app = express();
        app.all('*', function (req, res) {
            var body = '';
            req.on('data', function (chunk) {
                body += chunk;
            });
            req.on('end', function () {
                res.send(serializeRequest(req, body));
            });
        });
        server = app.listen(PORT, function () {
            done();
        });
        process.platform === oldPlatform;
        if (process.env.REQUEST_AS_CURL_COVERAGE) {
            delete require.cache[require.resolve('../lib-cov/escapeCliCommand')];
        } else {
            delete require.cache[require.resolve('../lib/escapeCliCommand')];
        }
    });

    afterEach(function (done) {
        server.close();
        setTimeout(done, 20);
    });

    it('should be able to serialize http.request', function (done) {
        var req = httpRequest({
            hostname: HOSTNAME,
            port: PORT,
            path: PATH,
            method: 'POST'
        },
        BODY,
        function (error, response, expected) {
            expect(serializeRequest(req, BODY)).to.eql(expected);
            done();
        });
    });

    it('should be able to serialize http.request with empty body', function (done) {
        var req = httpRequest({
                hostname: HOSTNAME,
                port: PORT,
                path: PATH,
                method: 'GET'
            },
            null,
            function (error, response, expected) {
                expect(serializeRequest(req)).to.eql(expected);
                done();
            });
    });

    it('should be able to serialize PUT http.request', function (done) {
        var req = httpRequest({
                hostname: HOSTNAME,
                port: PORT,
                path: PATH,
                method: 'PUT'
            },
            null,
            function (error, response, expected) {
                expect(serializeRequest(req)).to.eql(expected);
                done();
            });
    });

    it('should be able to serialize empty request', function () {
        expect(serializeRequest()).to.eql('curl \'http://localhost\' --compressed');
    });

    it('should be able to serialize https request', function () {
        expect(serializeRequest({
            headers: {
                host: 'localhost:443'
            }
        })).to.eql('curl \'https://localhost:443\' --compressed');
    });

    it('should be able to serialize broken ClientRequest', function (done) {
        var req = new http.get('http://' + HOSTNAME + ':' + PORT + PATH, function () {
            delete req._header;
            expect(serializeRequest(req)).to.eql('curl \'http://localhost:54321/p/a/t/h?query=string\' --compressed');
            done();
        });
    });

    it('should be able to escape query for win32', function (done) {
        process.platform = 'win32';
        var req = httpRequest({
                hostname: HOSTNAME,
                port: PORT,
                path: PATH,
                method: 'POST'
            },
            BODY,
            function (error, response, expected) {
                expect(serializeRequest(req, BODY)).to.eql(expected);
                done();
            });
    });

    it('should be able to serialize npm request module', function (done) {
        var req = request('http://' + HOSTNAME + ':' + PORT + PATH,
        {
            method: 'POST',
            body: BODY
        },
        function (error, response, expected) {
            expect(serializeRequest(req.req, BODY)).to.eql(expected);
            done();
        });
    });

    it('should be able to serialize npm request module with data json', function (done) {
        var req = request('http://' + HOSTNAME + ':' + PORT + PATH,
        {
            method: 'POST',
            json: BODY_JSON
        },
        function (error, response, expected) {
            expect(serializeRequest(req.req, BODY_JSON)).to.eql(expected);
            done();
        });
    });
});


lib/escapeCliCommand.js
/*
 * Copyright (C) 2007, 2008 Apple Inc.  All rights reserved.
 * Copyright (C) 2008, 2009 Anthony Ricaud <rik@webkit.org>
 * Copyright (C) 2011 Google Inc. All rights reserved.
 * Copyright (C) 2013 Mikhail Davydov <i@azproduction.ru>
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *
 * 1.  Redistributions of source code must retain the above copyright
 *     notice, this list of conditions and the following disclaimer.
 * 2.  Redistributions in binary form must reproduce the above copyright
 *     notice, this list of conditions and the following disclaimer in the
 *     documentation and/or other materials provided with the distribution.
 * 3.  Neither the name of Apple Computer, Inc. ("Apple") nor the names of
 *     its contributors may be used to endorse or promote products derived
 *     from this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY APPLE AND ITS CONTRIBUTORS "AS IS" AND ANY
 * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL APPLE OR ITS CONTRIBUTORS BE LIABLE FOR ANY
 * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
 * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
 * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
 * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

function escapeStringWin(str) {
    /* Replace quote by double quote (but not by \') because it is
     recognized by both cmd.exe and MS Crt arguments parser.

     Replace % by '%' because it could be expanded to an environment
     variable value. So %% becomes '%''%'. Even if an env variable ''
     (2 doublequotes) is declared, the cmd.exe will not
     substitute it with its value.

     Replace each backslash with double backslash to make sure
     MS Crt arguments parser won't collapse them.

     Replace new line outside of quotes since cmd.exe doesn't let
     to do it inside.
     */
    return "\"" + str.replace(/"/g, "\"\"")
        .replace(/%/g, "\"%\"")
        .replace(/\\/g, "\\\\")
        .replace(/[\r\n]+/g, "\"^$&\"") + "\"";
}

function escapeCharacter(x) {
    var code = x.charCodeAt(0);
    if (code < 256) {
        // Add leading zero when needed to not care about the next character.
        return code < 16 ? "\\x0" + code.toString(16) : "\\x" + code.toString(16);
    }
    code = code.toString(16);
    return "\\u" + ("0000" + code).substr(code.length, 4);
}

function escapeStringPosix(str) {
    if (/[^\x20-\x7E]|\'/.test(str)) {
        // Use ANSI-C quoting syntax.
        return "$\'" + str.replace(/\\/g, "\\\\")
            .replace(/\'/g, "\\\'")
            .replace(/\n/g, "\\n")
            .replace(/\r/g, "\\r")
            .replace(/[^\x20-\x7E]/g, escapeCharacter) + "'";
    } else {
        // Use single quote syntax.
        return "'" + str + "'";
    }
}

// cURL command expected to run on the same platform that DevTools run
// (it may be different from the inspected page platform).
module.exports = function (str) {
    return (process.platform === 'win32' ? escapeStringWin : escapeStringPosix)(str);
};


lib/index.js
/*
 * Copyright (C) 2007, 2008 Apple Inc.  All rights reserved.
 * Copyright (C) 2008, 2009 Anthony Ricaud <rik@webkit.org>
 * Copyright (C) 2011 Google Inc. All rights reserved.
 * Copyright (C) 2013 Mikhail Davydov <i@azproduction.ru>
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *
 * 1.  Redistributions of source code must retain the above copyright
 *     notice, this list of conditions and the following disclaimer.
 * 2.  Redistributions in binary form must reproduce the above copyright
 *     notice, this list of conditions and the following disclaimer in the
 *     documentation and/or other materials provided with the distribution.
 * 3.  Neither the name of Apple Computer, Inc. ("Apple") nor the names of
 *     its contributors may be used to endorse or promote products derived
 *     from this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY APPLE AND ITS CONTRIBUTORS "AS IS" AND ANY
 * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL APPLE OR ITS CONTRIBUTORS BE LIABLE FOR ANY
 * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
 * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
 * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
 * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */
var url = require('url'),
    http = require('http'),
    ClientRequest  = http.ClientRequest,
    escapeString = require('./escapeCliCommand');

var reTextData = /^application\/(x-www-form-urlencoded|json)/,
    reHttps = /443$/;

/**
 *
 * @param {Object} request
 * @param {String} request.url
 * @param {String} request.method
 * @param {Object} request.headers
 * @param {String|Object} [request.body]
 * @return {Array}
 */
function getUrl(request) {
    var host = request.headers.host || 'localhost';
    var protocol = reHttps.test(host) ? 'https' : 'http';
    var path = request.url;

    var queryUrl = url.format({
        protocol: protocol,
        host: host
    }) + (path ? path : '');

    return [escapeString(queryUrl).replace(/[[{}\]]/g, '\\$&')];
}

/**
 *
 * @param {Object} request
 * @param {String} request.url
 * @param {String} request.method
 * @param {Object} request.headers
 * @param {String|Object} [request.body]
 * @return {Array}
 */
function getData(request) {
    var data = [];
    var requestContentType = request.headers['content-type'];
    var body = request.body;

    if (!body) {
        return data;
    }

    body = typeof body === 'object' ? JSON.stringify(body) : body;

    if (requestContentType && reTextData.test(requestContentType)) {
        data.push('--data');
    } else {
        data.push('--data-binary');
    }

    data.push(escapeString(body));

    return data;
}

/**
 *
 * @param {Object} request
 * @param {String} request.url
 * @param {String} request.method
 * @param {Object} request.headers
 * @param {String|Object} [request.body]
 * @param {Object} ignoredHeaders
 * @return {Array}
 */
function getHeaders(request, ignoredHeaders) {
    var headers = [];
    Object.keys(request.headers).forEach(function (headerName) {
        var headerValue = request.headers[headerName];
        // Translate SPDY v3 headers to HTTP headers.
        headerName = headerName.replace(/^:/, '');

        if (headerName.toLowerCase() in ignoredHeaders) {
            return;
        }
        headers.push('-H');
        headers.push(escapeString(headerName + ': ' + headerValue));
    });

    return headers;
}

function normalizeHeaders(request) {
    var requestHeaders;

    if (request instanceof ClientRequest) {
        // POST / HTTP/1.1\r\nHost: localhost:54321\r\nConnection: keep-alive\r\nTransfer-Encoding: chunked\r\n\r\n
        if (typeof request._header === 'string') {
            requestHeaders = request._header
                .split('\r\n')
                .slice(1, -2)
                .reduce(function (headers, headerString) {
                    var parts = headerString.split(':'),
                        header = parts.shift().trim().toLowerCase(),
                        value = parts.join(':').trim();

                    headers[header] = value;
                    return headers;
                }, {});
        } else {
            requestHeaders = request._headers;
        }
    } else {
        requestHeaders = request.headers;
    }
    return requestHeaders || {};
}

function normalizeUrl(request) {
    if (request instanceof ClientRequest) {
        return request.path;
    }

    return request.url;
}

function normalize(request, body) {
    request = request || {};

    return {
        method: request.method || 'GET',
        url: normalizeUrl(request),
        headers: normalizeHeaders(request),
        body: body
    };
}

/**
 *
 * @param {Object} request
 * @param {String} request.url
 * @param {String} request.method
 * @param {Object} request.headers
 * @param {String|Object} [request.body]
 *
 * @returns {string}
 */
function serializeHttpRequestAsCurl(request) {
    var command = ['curl'];
    // These headers are derived from URL (except 'version') and would be added by cURL anyway.
    var ignoredHeaders = {'host': 1, 'method': 1, 'path': 1, 'scheme': 1, 'version': 1};
    var inferredMethod = 'GET';

    // URL
    command = command.concat(getUrl(request));

    // DATA
    var data = getData(request);
    if (data.length) {
        inferredMethod = 'POST';
        ignoredHeaders['content-length'] = 1;
    }

    // METHOD
    if (request.method !== inferredMethod) {
        command.push('-X', request.method);
    }

    // HEADERS & DATA
    command = command.concat(getHeaders(request, ignoredHeaders))
        .concat(data)
        .concat('--compressed');

    return command.join(' ');
}

/**
 * Serializes http.ClientRequest as curl(1) command string
 *
 * @param {http.ClientRequest|http.IncomingMessage|*} request
 * @param {string|object} [body]
 *
 * @returns {string}
 */
module.exports = function (request, body) {
    return serializeHttpRequestAsCurl(normalize(request, body));
};

module.exports.serialize = serializeHttpRequestAsCurl;
